src/network.rs

```diff
@@ -0,0 +1,117 @@
+use super::daemon::Daemon;
+use super::error::Error;
+use super::logger;
+use super::logger::Interface;
+use std::fmt;
+use std::fmt::{Display, Formatter};
+use std::io::Write;
+use std::io::{BufRead, BufReader};
+use std::os::unix::net::UnixListener;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::thread;
+
+#[derive(PartialEq, Debug)]
+pub enum Packet {
+    Hello(String),
+    HelloResponse(String, u32),
+    Unknown,
+}
+
+pub fn parse_packet(packet: &String) -> Result<Packet, Error> {
+    let mut packet_split = packet.split("|");
+    let packet_type = packet_split.next().unwrap_or("?");
+    let packet_data = packet_split.next().unwrap_or("?");
+    match packet_type {
+        "0" => Ok(Packet::Hello(packet_data.to_string())),
+        "1" => Ok(Packet::HelloResponse(
+            packet_data.to_string(),
+            packet_split.next().unwrap().parse::<u32>().unwrap_or(0),
+        )),
+        _ => Ok(Packet::Unknown),
+    }
+}
+
+impl Display for Packet {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            Packet::Hello(data) => write!(f, "0|{}", data),
+            Packet::HelloResponse(data, version) => write!(f, "1|{}|{}", data, version),
+            Packet::Unknown => write!(f, ""),
+        }
+    }
+}
+
+fn log_to_daemon(daemon: &Arc<Mutex<Daemon>>, message: &str, severity: logger::Severity) {
+    let mut daemon = daemon.lock().unwrap();
+    daemon.logger.log(
+        message,
+        severity,
+    );
+}
+
+pub fn listen(path: &'static str, c_daemon_mutex: Arc<Mutex<Daemon>>) {
+    thread::spawn(move || {
+        // Get rid of the old sock
+        std::fs::remove_file(path).ok();
+
+        // Try to handle sock connections then
+        let listener = match UnixListener::bind(path) {
+            Ok(listener) => listener,
+            Err(e) => {
+                log_to_daemon(&c_daemon_mutex, &format!("Failed to bind to {}: {}", path, e), logger::Severity::Error);
+                return;
+            }
+        };
+
+        // Spawn a new thread to listen for commands
+        thread::spawn(move || {
```

Comment:
Wow, this is mega smart, now I see why threads are important here. Zero blocking, smart

Reply:
Yup it made stuff complicated but overall it should be efficient.